session/index.js

```diff
@@ -0,0 +1,171 @@
+      if (this.sessionInfo.autoTrack) {
+        this.startAutoTracking();
+      } else if (
+        Object.keys(this.sessionInfo).length > 1 &&
+        !this.sessionInfo.manuallyTrackSession
+      ) {
+        /**
+         * Use case:
+         * By default user session is enabled which means storage will have session data.
+         * In case user wanted to opt out and set auto track to false through load option,
+         * clear stored session info.
+         */
+        this.end();
+      }
+    } catch (e) {
+      handleError(e);
+    }
+  }
+
+  /**
+   * A function to validate current session and return true/false depending on that
+   * @param {number} timestamp
+   * @returns boolean
+   */
+  isValidSession(timestamp) {
+    return timestamp <= this.sessionInfo.expiresAt;
+  }
+
+  /**
+   * A function to check for existing session details and depending on that create a new session.
+   */
+  startAutoTracking() {
+    const timestamp = Date.now();
+    if (!this.isValidSession(timestamp)) {
+      this.sessionInfo.id = timestamp.toString(); // set the current timestamp
+      this.sessionInfo.expiresAt = timestamp + this.timeout; // set the expiry time of the session
+      this.sessionInfo.sessionStart = true;
+      this.sessionInfo.autoTrack = true;
+    }
+    this.storage.setSessionInfo(this.sessionInfo);
+  }
+
+  /**
+   * A public method to start a session
+   * @param {string} sessionId     session identifier
+   * @returns
+   */
+  start(sessionId) {
+    if (this.sessionInfo.manuallyTrackSession) return;
+    if (sessionId && typeof sessionId !== 'string') {
+      logger.error(`[Session]:: "sessionId" should be in string format`);
+      return;
+    }
+    this.sessionInfo = {
+      id: sessionId || Date.now().toString(),
+      sessionStart: true,
+      manuallyTrackSession: true,
+      autoTrack: false,
+    };
+    this.storage.setSessionInfo(this.sessionInfo);
+  }
+
+  /**
+   * A public method to end an ongoing session.
+   */
+  end() {
+    this.sessionInfo = {};
+    this.storage.removeSessionInfo();
+  }
+
+  /**
+   * A function get ongoing sessionId.
+   */
+  getSession() {
+    const session = {};
+    if (this.sessionInfo.autoTrack || this.sessionInfo.manuallyTrackSession) {
+      const timestamp = Date.now();
+      if (!this.sessionInfo.manuallyTrackSession) {
+        if (!this.isValidSession(timestamp)) this.startAutoTracking();
+        else {
+          this.sessionInfo.expiresAt = timestamp + this.timeout; // set the expiry time of the session
+        }
+      }
+      if (this.sessionInfo.sessionStart) {
+        this.sessionInfo.sessionStart = false;
+        session.sessionStart = true;
+      }
+      this.storage.setSessionInfo(this.sessionInfo);
+      session.sessionId = this.sessionInfo.id;
+    }
+    return session;
+  }
+
+  reset() {
+    const { manuallyTrackSession, autoTrack } = this.sessionInfo;
+    this.sessionInfo = {};
```

Comment:
This is not required if you address one of the comments above.

Reply:
Addressed